docs/blog/posts/constitutional-ai-or-validation.md

```diff
@@ -0,0 +1,221 @@
+---
+draft: False 
+date: 2023-10-17
+tags:
+    - pydantic
+    - validation
+    - guardrails 
+    - constitutional ai
+---
+
+# Constitutional AI is just Probabilistic Validation
+
+Earlier this year, there was a lot of discussion about Constitutional AI and its role in guiding models. There was also talk about the need for guard rails and special systems to ensure safe outputs. However, I found myself unimpressed. Why create a new term when we already have one that describes the concept? That term is "validation," and now we can do it probabilistically.
+
+Tools like Constitutional AI address shortcomings by using AI feedback to evaluate outputs. In this approach, an AI system provides a set of principles for making judgments about generated text. These principles guide the model's behavior. Essentially, it's a form of validation. Instead of manually checking against a set of rules, we now have a model that can perform the validation. For example, a content moderation system could use a model to automatically ban certain keywords. Previously, an error might have been thrown upon detection, but now we have a self-correcting system.
+
+This post will explore how we can achieve this using Pydantic and Instructor without introducing new standards or terms for validation.
+
+## Validation in Software 1.0
+
+Common types of validation in Software 1.0 include:
+
+-  Type checking
+-  Range checking
+-  Length checking
+
+Here's an example model that uses Pydantic to validate the range and length of a string, taken directly from the Pydantic documentation:
+
+```python
+from pydantic import BaseModel, ValidationError, field_validator
+
+class UserModel(BaseModel):
+    id: int
+    name: str
+
+    @field_validator('name')
+    @classmethod
+    def name_must_contain_space(cls, v: str) -> str:
+        if ' ' not in v:
+            raise ValueError('must contain a space')
+        return v.title()
```

Comment:
I know this is an example but it'd be interesting if you used more PEP 593 validated types:

```python
from typing import Annotated

from pydantic import AfterValidator, BaseModel

def name_must_contain_space(cls, v: str) -> str:
        if ' ' not in v:
            raise ValueError('must contain a space')
        return v.title()

Name = Annotated[str, AfterValidator(name_must_contain_space)]
"""A name. It must not contain a space!"""

class UserModel(BaseModel):
    id: int
    name: Name
```

Reply:
yeah my docs end u using that, 

```
AfterValidator(llm_validator("..."))
```

Do you think its confusing to do both?

Reply:
No it's not, just that I find now I almost always want to use the `Annotated` version